### PR TITLE
mfd: Add Raspberry Pi Sense HAT core driver

### PR DESCRIPTION
**Pull request by XECDesign**
joystick: Add Raspberry Pi Sense HAT joystick driver
fbdev: Add Raspberry Pi Sense HAT framebuffer driver
configs: Enable Raspberry Pi Sense HAT modules
overlays: Provide a Raspberry Pi Sense HAT overlay for older Raspberry Pi models

### CODE REVIEW

File: drivers/input/joystick/rpisense-js.c [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	rpisense_js->keys_dev->keycodemax = ARRAY_SIZE(keymap);
+
+	result = input_register_device(rpisense_js->keys_dev);
+
+	if (result) {
+		dev_err(&pdev->dev, "Could not register input device.\n");
+		goto err_keys_alloc;
+	}
+
+	rpisense_js->keys_desc = gpiod_get(rpisense->dev, "keys-int", GPIOD_IN);
```
Comment (pelwell) [inline drivers/input/joystick/rpisense-js.c]: If you use the managed version (devm_gpiod_get) then you don't need to call gpiod_put.

**Comment by pelwell** [comment]
Can you add an entry to the overlays/README file?

**Comment by XECDesign** [comment]
I guess I shouldn't squash the changes until you and Dom are okay to merge it, to keep the changes easier to track

**Comment by pelwell** [comment]
I think at this stage it would help if you did squash. Dom hasn't looked at it yet, and it would save time if he just saw the "final" version.

**Comment by XECDesign** [comment]
done

**Comment by XECDesign** [comment]
@notro If you get a chance, would you be willing to take a look at the framebuffer section? The documentation assumes more knowledge of traditional graphics hw than I possess, so there's a lot of guesswork in the code.

**Comment by pelwell** [comment]
It looks OK to me, apart from the commit message which is enormous. Either rewrite it completely or just add a summary title line, but make sure you put a blank line after the first line.

**Comment by notro** [comment]
``` diff
diff --git a/drivers/video/fbdev/rpisense-fb.c b/drivers/video/fbdev/rpisense-fb.c

+static void *rvmalloc(unsigned long size)
+{
+   void *mem;
+   unsigned long adr;
+   size = PAGE_ALIGN(size);
+   mem = vmalloc_32(size);
+   if (!mem)
+       return NULL;
+
+   memset(mem, 0, size); /* Clear the ram out, no junk to the user */
+   adr = (unsigned long) mem;
+   while (size > 0) {
+       SetPageReserved(vmalloc_to_page((void *)adr));
+       adr += PAGE_SIZE;
+       size -= PAGE_SIZE;
+   }
+
+   return mem;
+}
```

I have seen this type of code in out-of-tree fb drivers, but not in mainline.
In my fbtft drivers I just use vzalloc(size) to allocate the framebuffer memory, no SetPageReserved().

**Comment by notro** [comment]
@XECDesign Some kernel coding style issues you might wanna look into: https://gist.github.com/notro/c73b9342e52f89494311

**Comment by XECDesign** [comment]
Ah, didn't know about checkpatch script. Thanks.

**Comment by XECDesign** [comment]
fixed

**Comment by P33M** [comment]
@pelwell ping - ok to merge?

**Comment by pelwell** [comment]
Is it picky to ask for the overlay makefile addition to also be in alphabetical order, like the others?

**Comment by XECDesign** [comment]
Done

**Pull request merged**